google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionPoolTest.java

```diff
@@ -1551,6 +1588,193 @@ public void testSessionMetrics() throws Exception {
+    if (minSessions == 1) {
+      options =
+          SessionPoolOptions.newBuilder()
+              .setMinSessions(1)
+              .setMaxSessions(3)
+              .setMaxIdleSessions(0)
+              .setInitialWaitForSessionTimeoutMillis(50L)
+              .build();
+      FakeClock clock = new FakeClock();
+      clock.currentTimeMillis = System.currentTimeMillis();
+
+      InMemoryMetricReader inMemoryMetricReader = InMemoryMetricReader.create();
+
+      SdkMeterProvider sdkMeterProvider =
+          SdkMeterProvider.builder().registerMetricReader(inMemoryMetricReader).build();
+
+      OpenTelemetry openTelemetry =
+          OpenTelemetrySdk.builder().setMeterProvider(sdkMeterProvider).build();
+
+      setupMockSessionCreation();
+
+      AttributesBuilder attributesBuilder = Attributes.builder();
+      attributesBuilder.put("client_id", "testClient");
+      attributesBuilder.put("database", "testDb");
+      attributesBuilder.put("instance_id", "test_instance");
+      attributesBuilder.put("library_version", "test_version");
+
+      pool =
+          createPool(
+              clock,
+              Metrics.getMetricRegistry(),
+              SPANNER_DEFAULT_LABEL_VALUES,
+              openTelemetry,
+              attributesBuilder.build());
+      PooledSessionFuture session1 = pool.getSession();
+      PooledSessionFuture session2 = pool.getSession();
+      session1.get();
+      session2.get();
+
+      Collection<MetricData> metricDataCollection = inMemoryMetricReader.collectAllMetrics();
+
+      // Acquired sessions are 2.
+      verifyMetricData(metricDataCollection, NUM_ACQUIRED_SESSIONS_NAME, 1, 2L);
+
+      // Max in use session are 2.
+      verifyMetricData(metricDataCollection, MAX_IN_USE_SESSIONS_NAME, 1, 2D);
+
+      // Max Allowed sessions should be 3
+      verifyMetricData(metricDataCollection, MAX_ALLOWED_SESSIONS_NAME, 1, 3D);
+
+      // Released sessions should be 0
+      verifyMetricData(metricDataCollection, NUM_RELEASED_SESSIONS_NAME, 1, 0L);
+
+      // Num sessions in pool
+      verifyMetricData(
+          metricDataCollection, NUM_SESSIONS_IN_POOL_NAME, 1, NUM_SESSIONS_IN_USE_NAME, 2);
+
+      PooledSessionFuture session3 = pool.getSession();
+      session3.get();
+
+      final CountDownLatch latch = new CountDownLatch(1);
+      // Try asynchronously to take another session. This attempt should time out.
+      Future<Void> fut =
+          executor.submit(
+              () -> {
+                latch.countDown();
+                Session session = pool.getSession();
+                session.close();
+                return null;
+              });
+      // Wait until the background thread is actually waiting for a session.
+      latch.await();
+      // Wait until the request has timed out.
+      int waitCount = 0;
+      while (pool.getNumWaiterTimeouts() == 0L && waitCount < 1000) {
+        Thread.sleep(5L);
+        waitCount++;
+      }
```

Comment:
(Just checking) Would it be possible to replace this busy-wait with some kind of test setup that advances the `FakeClock` and causes a timeout that way?